src/network/net_addons.cc

```diff
@@ -160,6 +161,186 @@ void append_multiline_message(std::string& send, const std::string& message) {
+		const size_t id = ++cmd_id_counter_;
+
+		{
+			MutexLock m(queue_mutex_);
+			queue_.emplace_back(socket, buffer, length);
+			queue_.back().id_ = id;
+		}
+
+		const uint32_t start_time = SDL_GetTicks();
+		for (;;) {
+			{
+				MutexLock m(result_mutex_);
+				auto it = return_values_.find(id);
+				if (it != return_values_.end()) {
+					int result = it->second;
+					return_values_.erase(it);
+					hangup_fn_(SDL_GetTicks() - start_time, true);
+					return result;
+				}
+			}
+
+			hangup_fn_(SDL_GetTicks() - start_time, false);
+			SDL_Delay(1);
+		}
+	}
+
+	static void run(AsyncIOWrapper* io) {  // Runs in the separate thread `io->thread_`.
+		while (io->alive_) {
+			QueuedCommand cmd;
+			{
+				MutexLock m(io->queue_mutex_);
+				if (!io->queue_.empty()) {
+					cmd = io->queue_.front();
+					assert(cmd.id_ != 0);
+					io->queue_.pop_front();
+				}
+			}
+
+			if (cmd.id_ != 0) {
+				assert(io->return_values_.count(cmd.id_) == 0);
+				int result;
+				switch (cmd.type_) {
+				case QueuedCommand::Type::kWrite:
+					result = portable_write(cmd.socket_, cmd.buffer_.write, cmd.length_);
+					break;
+
+				case QueuedCommand::Type::kRead: {
+					std::unique_ptr<char[]> temp_buffer(new char[cmd.length_]);
+					result = portable_read(cmd.socket_, temp_buffer.get(), cmd.length_);
+					if (io->alive_) {
+						memcpy(cmd.buffer_.read, temp_buffer.get(), cmd.length_);
+					}
+					break;
+				}
+
+				case QueuedCommand::Type::kReadLine: {
+					std::string temp_readout;
+					char c;
+					for (;;) {
+						result = portable_read(cmd.socket_, &c, 1);
+						if (result != 1 || c == '\n') {
+							break;
+						}
+						temp_readout += c;
+					}
+					if (io->alive_) {
+						*cmd.buffer_.read_line = temp_readout;
+					}
+					break;
+				}
+				}
+
+				MutexLock m(io->result_mutex_);
+				io->return_values_.emplace(cmd.id_, result);
+			} else {
+				SDL_Delay(1);
+			}
+		}
+
+		verb_log_info("Async network IO thread ending");
+		io->thread_.detach();
+		MutexLock m(MutexLock::ID::kLog);
+		living_async_wrappers.erase(io);
+		delete io;
+	}
+
+	const MutexLock::ID queue_mutex_;
+	const MutexLock::ID result_mutex_;
+	HangupFn hangup_fn_;
+	volatile bool alive_{true};
+
+	size_t cmd_id_counter_{0U};
+	std::deque<QueuedCommand> queue_;
+	std::map<size_t, int> return_values_;
+
+	std::thread thread_;
+};
+
+void cleanup_abandoned_hung_threads() {
+	const uint32_t start = SDL_GetTicks();
+	constexpr uint32_t kKillTimeout = 5000;
+	do {
+		if (living_async_wrappers.empty()) {
+			return;
+		}
+		SDL_Delay(1);
+	} while (SDL_GetTicks() - start < kKillTimeout);
+
+	// This is called during shutdown, the logger might no longer exist
+	std::cout << "FATAL: Unable to clean up " << living_async_wrappers.size()
+	          << " hung network threads, killing Widelands.\n";
+
+	// TODO(Nordfriese): There must be a better way to do interrupt a blocked read() call…
+	std::terminate();
```

Comment:
I did not get here, the thread was already terminated after closing the message box.